src/main/scala/com/microsoft/hyperspace/actions/CreateActionBase.scala

```diff
@@ -124,6 +128,10 @@ private[actions] abstract class CreateActionBase(dataManager: IndexDataManager)
   protected def write(spark: SparkSession, df: DataFrame, indexConfig: IndexConfig): Unit = {
     val numBuckets = numBucketsForIndex(spark)
 
+    // If adding lineage, make sure a unique file id is already assigned to each source data file.
```

Comment:
Yes, to build fileIdTracker, `def getIndexLogEntry` should be called earlier than `def write`. 
In Action, `before` and `end` call `logEntry`, so there's no problem.
But there is no assurance for that, thus this is the minimum check for it.

And seems  we could remove "indexLineageEnabled(spark) now.

Reply:
> Is this assert still valid since we are assigning ids regardless?

This is the very last step before doing the actual broadcast join to populate lineage column and we wanna make sure the ids DF is actually populated with the ids.

> we could remove "indexLineageEnabled(spark) now

We actually need it. When creating an index, we look at the lineage configuration flag to decide if lineage should be enabled or not. When refreshing, we should look at previous index logEntry. Keeping the function and overriding it in RefreshActionBase is a clean/clear way to capture this.

Reply:
This assert still confuses me. In refresh, `fileIdTracker` is from the previous log entry, meaning it may contain both source and index. Then does `fileIdTracker.getFileToIdMap.size >= df.inputFiles.length` guarantee "make sure a unique file id is already assigned to each source data file"? If you want to do what the comment says, you have to iterate `df.inputFiles` and make sure each one is in the `fileIdTracker.getFileToIdMap`.

Also, I am not sure why `!hasLineage(spark)` is needed since `fileIdTracker` behavior does not depend on lineage (unless I missed some parts of code)

Reply:
@imback82 - Those are actually good points and yes the assert now should be revised after adding id generation for index files. 
Again our test code already checks this and with the way bein(),op(),end() are called in our Actions there should be no problem here.
This assert was originally added because of a comment by @sezruby . Now that things are changed (i.e. we generate ids for index files and retain all ids during all refresh modes), I think we should drop this assert. @sezruby  WDYT?

Reply:
+1 for dropping if not needed.

Reply:
ok then let's remove this and how about adding row count check after join query? Just in case? @pirz